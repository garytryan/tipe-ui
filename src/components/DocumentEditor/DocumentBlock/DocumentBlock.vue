<template>
  <div
    :data-tipe-ui="$options.name"
    :class="rootStyleObject"
    @mouseover="hover = true"
    @mouseout="hover = false"
  >
    <div class="margin">
      <tipe-document-block-margin
        v-if="hover"
        :options="options"
        :waiting="block.waiting"
        :disabled="block.disabled"
      />
    </div>
    <div class="header">
      <tipe-document-block-header
        v-bind="block"
        @change="value => $emit('change', value)"
      />
    </div>
    <div class="body">
      <tipe-document-block-value-input
        :status="block.status"
        :value="block.value"
        :waiting="block.waiting"
        :disabled="block.disabled"
        :type="block.type"
        @change="value => $emit('change', value)"
      />
    </div>
    <div class="footer">
      <tipe-document-block-message
        v-if="block.status != ''"
        :success-message="block.errorMessage"
        :warning-message="block.errorMessage"
        :error-message="block.errorMessage"
        :status="block.status"
      />
    </div>
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import documentBlockShape from '@/types/DocumentBlock'
import blockOptionShape from '@/types/BlockOption'
import TipeDocumentBlockValueInput from './DocumentBlockValueInput'
import TipeDocumentBlockHeader from './DocumentBlockHeader'
import TipeDocumentBlockMargin from './DocumentBlockMargin'
import TipeDocumentBlockMessage from './DocumentBlockMessage'

export default {
  name: 'TipeDocumentBlock',
  components: {
    TipeDocumentBlockValueInput,
    TipeDocumentBlockHeader,
    TipeDocumentBlockMargin,
    TipeDocumentBlockMessage
  },
  props: {
    options: vueTypes.arrayOf(vueTypes.shape(blockOptionShape)),
    block: vueTypes.shape(documentBlockShape)
  },
  data() {
    return {
      hover: false
    }
  },
  computed: {
    rootStyleObject() {
      return {
        [`status-${this.block.status}`]: this.block.status,
        waiting: this.block.waiting,
        disabled: this.block.disabled
      }
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentBlock'] {
  display: grid;
  grid-template-areas:
    'margin header'
    'margin body'
    'margin footer';
  grid-template-columns: 4.25rem 1fr;
  grid-template-rows: auto auto auto;
  border-radius: 0.1875rem;
  box-sizing: border-box;
  width: 100%;
  min-width: 20rem;
}

.margin {
  grid-area: margin;
}

.header {
  grid-area: header;
}

.body {
  grid-area: body;
}

.footer {
  grid-area: footer;
}

[data-tipe-ui='TipeDocumentBlock'].waiting {
  cursor: wait;
}

[data-tipe-ui='TipeDocumentBlock'].disabled {
  cursor: not-allowed;
}
</style>
