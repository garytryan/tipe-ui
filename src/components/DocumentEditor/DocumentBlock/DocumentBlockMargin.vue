<template>
  <div
    :data-tipe-ui="$options.name"
    :class="rootStyleObject"
  >
    <div class="grab-handle">
      <tipe-icon icon="options" />
    </div>
    <tipe-select-block
      :waiting="waiting"
      :disabled="disabled"
      :options="options"
    />
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import TipeIcon from '@/components/Icon'
import interfaces from '@tipe/tipe-interfaces'
import TipeSelectBlock from './SelectBlock'

export default {
  name: 'TipeDocumentBlockMargin',
  components: {
    TipeIcon,
    TipeSelectBlock
  },
  props: {
    options: vueTypes.arrayOf(vueTypes.shape(interfaces.blockOption)),
    waiting: vueTypes.bool.def(false),
    disabled: vueTypes.bool.def(false)
  },
  computed: {
    rootStyleObject() {
      return {
        waiting: this.waiting,
        disabled: this.disabled
      }
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentBlockMargin'] {
  display: flex;
  flex-direction: row;
  align-items: flex-start;
  justify-content: flex-start;
  height: 100%;
  width: 100%;
}

.grab-handle {
  display: grid;
  height: 1.75rem;
  width: 1rem;
  align-items: center;
  justify-content: center;
  cursor: grab;
}

[data-tipe-ui='TipeDocumentBlock'].waiting {
  cursor: wait;
}

[data-tipe-ui='TipeDocumentBlock'].disabled {
  cursor: not-allowed;
}
</style>
