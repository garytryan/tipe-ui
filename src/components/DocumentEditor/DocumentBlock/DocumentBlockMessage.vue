<template>
  <div
    :data-tipe-ui="$options.name"
    :class="rootStyleObject"
  >
    {{ message }}
  </div>
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeDocumentBlockMessage',
  props: {
    status: vueTypes.oneOf(['success', 'warning', 'error', '']).def(''),
    successMessage: vueTypes.string.def(''),
    warningMessage: vueTypes.string.def(''),
    errorMessage: vueTypes.string.def('')
  },
  computed: {
    rootStyleObject() {
      return {
        [`status-${this.status}`]: this.status,
        waiting: this.waiting,
        disabled: this.disabled
      }
    },
    message() {
      switch (this.status) {
        case 'success':
          return this.successMessage
        case 'warning':
          return this.warningMessage
        case 'error':
          return this.errorMessage
        default:
          return ''
      }
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentBlockMessage'] {
  height: 1.5rem;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  font-size: 0.8125rem;
  white-space: nowrap;
  overflow: hidden;
}

[data-tipe-ui='TipeDocumentBlockMessage'].status-success {
  color: var(--success);
  border-bottom: 1px solid var(--success);
}

[data-tipe-ui='TipeDocumentBlockMessage'].status-warning {
  color: var(--warning);
  border-bottom: 1px solid var(--warning);
}

[data-tipe-ui='TipeDocumentBlockMessage'].status-error {
  color: var(--error);
  border-bottom: 1px solid var(--error);
}
</style>
