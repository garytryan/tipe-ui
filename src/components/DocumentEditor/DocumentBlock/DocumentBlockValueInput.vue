<template>
  <div :data-tipe-ui="$options.name">
    <component
      :status="status"
      :value="value"
      :waiting="waiting"
      :disabled="disabled"
      :is="is"
      size="medium"
      @change="onChange"
    />
  </div>
</template>

<script>
import documentBlockShape from '@/types/DocumentBlock'
import TipeSwitch from '@/components/Switch'

export default {
  name: 'TipeDocumentBlock',
  components: {
    TipeSwitch
  },
  props: documentBlockShape,
  computed: {
    is() {
      return 'tipe-switch'
    }
  },
  methods: {
    onChange(event) {
      this.$emit('change', { value: event.target.value })
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentBlock'] {
  padding: 1.5rem 0 1rem 0;
}
</style>
