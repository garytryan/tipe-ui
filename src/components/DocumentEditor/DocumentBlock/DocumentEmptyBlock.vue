<template>
  <div
    :data-tipe-ui="$options.name"
  >
    <div class="margin">
      <tipe-document-block-margin
        :options="options"
      />
    </div>
    <div class="body">
      Please add a block.
    </div>
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import interfaces from '@tipe/tipe-interfaces'
import TipeDocumentBlockMargin from './DocumentBlockMargin'

export default {
  name: 'TipeDocumentEmptyBlock',
  components: {
    TipeDocumentBlockMargin
  },
  props: {
    options: vueTypes.arrayOf(vueTypes.shape(interfaces.blockOption))
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentEmptyBlock'] {
  display: grid;
  grid-template-areas: 'margin body';
  grid-template-columns: 4.25rem 1fr;
  grid-template-rows: auto;
  border-radius: 0.1875rem;
  box-sizing: border-box;
  width: 100%;
  min-width: 20rem;
}

.margin {
  grid-area: margin;
}

.body {
  grid-area: body;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  color: var(--midPurple);
}
</style>
