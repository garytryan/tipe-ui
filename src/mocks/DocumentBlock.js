import faker from 'faker'
import { getBlockTypes, getBlockLabel } from '@/libs/blocks'
import documentBlockValidationMock from '@/mocks/DocumentBlockValidation'

export default documentBlock => {
  const type = getBlockTypes()[
    faker.random.number({ min: 0, max: getBlockTypes().length - 1 })
  ]

  return {
    type,
    label: getBlockLabel(type),
    value: '',
    name: faker.lorem.word(),
    apiId: faker.random.word(),
    description: faker.lorem.sentence(),
    status: ['success', 'warning', 'error', undefined][
      faker.random.number({ min: 0, max: 3 })
    ],
    successMessage: faker.lorem.sentence(),
    errorMessage: faker.lorem.sentence(),
    warningMessage: faker.lorem.sentence(),
    waiting: faker.random.boolean(),
    disabled: faker.random.boolean(),
    validation: documentBlockValidationMock(),
    ...documentBlock
  }
}
