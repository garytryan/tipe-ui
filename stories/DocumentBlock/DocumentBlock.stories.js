import { storiesOf } from '@storybook/vue'
import TipeDocumentBlock from '@/components/DocumentEditor/DocumentBlock'
import { mocks } from '@tipe/tipe-test-utils'

storiesOf('DocumentBlock', module)
  .add('status(undefined)', () => ({
    components: { TipeDocumentBlock },
    data() {
      return {
        options: mocks.createManyMocks(mocks.blockOption, 3),
        block: mocks.documentBlock({
          value: true,
          status: '',
          waiting: false,
          disabled: false
        })
      }
    },
    template: '<tipe-document-block :block="block" :options="options" />'
  }))
  .add('status(success)', () => ({
    components: { TipeDocumentBlock },
    data() {
      return {
        options: mocks.createManyMocks(mocks.blockOption, 3),
        block: mocks.documentBlock({
          value: true,
          status: 'success',
          waiting: false,
          disabled: false
        })
      }
    },
    template: '<tipe-document-block :block="block" :options="options" />'
  }))
  .add('status(warning)', () => ({
    components: { TipeDocumentBlock },
    data() {
      return {
        options: mocks.createManyMocks(mocks.blockOption, 3),
        block: mocks.documentBlock({
          value: true,
          status: 'warning',
          waiting: false,
          disabled: false
        })
      }
    },
    template: '<tipe-document-block :block="block" :options="options" />'
  }))
  .add('status(error)', () => ({
    components: { TipeDocumentBlock },
    data() {
      return {
        options: mocks.createManyMocks(mocks.blockOption, 3),
        block: mocks.documentBlock({
          value: true,
          status: 'error',
          waiting: false,
          disabled: false
        })
      }
    },
    template: '<tipe-document-block :block="block" :options="options" />'
  }))
  .add('waiting', () => ({
    components: { TipeDocumentBlock },
    data() {
      return {
        options: mocks.createManyMocks(mocks.blockOption, 3),
        block: mocks.documentBlock({
          value: true,
          status: '',
          waiting: true,
          disabled: false
        })
      }
    },
    template: '<tipe-document-block :block="block" :options="options" />'
  }))
  .add('disabled', () => ({
    components: { TipeDocumentBlock },
    data() {
      return {
        options: mocks.createManyMocks(mocks.blockOption, 3),
        block: mocks.documentBlock({
          value: true,
          status: '',
          waiting: false,
          disabled: true,
          name: ''
        })
      }
    },
    template: '<tipe-document-block :block="block" :options="options" />'
  }))
