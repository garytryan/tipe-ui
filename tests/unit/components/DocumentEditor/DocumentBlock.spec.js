import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import DocumentBlock from '@/components/DocumentEditor/DocumentBlock'
import documentBlockFixture from '@/fixtures/DocumentBlock'
import documentBlockMock from '@/mocks/DocumentBlock'

describe('DocumentBlock', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const propsData = { block: documentBlockFixture() }
      const wrapper = shallowMount(DocumentBlock, { propsData })

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('has correct data-tipe-iu attibute', () => {
      const propsData = { block: documentBlockMock() }
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeDocumentBlock')
    })
  })

  describe(':props', () => {
    it(':status(undefined) - no status class ', () => {
      const propsData = { block: documentBlockMock() }
      delete propsData.block.status
      const wrapper = shallowMount(DocumentBlock, { propsData })
      expect(wrapper.classes()).not.toEqual(
        expect.arrayContaining([expect.stringMatching(/status-/)])
      )
    })

    it(':status(success) - has status-success class', () => {
      const propsData = { block: documentBlockMock({ status: 'success' }) }
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('status-success')
    })

    it(':status(error) - has status-error class', () => {
      const propsData = { block: documentBlockMock({ status: 'error' }) }
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('status-error')
    })

    it(':status(warning) - has status-warning class', () => {
      const propsData = { block: documentBlockMock({ status: 'error' }) }
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('status-error')
    })

    it(':waiting(false) - no waiting class', () => {
      const propsData = { block: documentBlockMock({ waiting: false }) }
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).not.toEqual(expect.arrayContaining(['waiting']))
    })

    it(':waiting(true) - has waiting class', () => {
      const propsData = { block: documentBlockMock({ waiting: true }) }
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('waiting')
    })
  })
})
