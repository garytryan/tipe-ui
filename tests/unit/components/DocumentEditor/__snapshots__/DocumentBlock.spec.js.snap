// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentBlock <template> matches previous snapshot 1`] = `
<div data-tipe-ui="TipeDocumentBlock" class="status-success">
  <div class="margin">
    <!---->
  </div>
  <div class="header">
    <TipeDocumentBlockHeader-stub></TipeDocumentBlockHeader-stub>
  </div>
  <div class="body">
    <TipeDocumentBlock-stub></TipeDocumentBlock-stub>
  </div>
  <div class="footer">
    <TipeDocumentBlockMessage-stub></TipeDocumentBlockMessage-stub>
  </div>
</div>
`;
